### PR TITLE
Closes #465

### LINKED ISSUE #465: [Broker Cache] FoCI Cache Interrogation API
Broker needs some mechanism to interrogate the cache to determine:
- "Is this client_id known to the cache?"

And some way to privately query for a `List` of FoCI `ICacheRecords` / `AccountRecords` (TBD)

### PR DESCRIPTION
**Pull request by iambmelt**
See https://github.com/AzureAD/microsoft-authentication-library-common-for-android/issues/465

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java [hunk 1]
```diff
@@ -750,6 +753,81 @@ public AccountDeletionRecord removeAccount(@Nullable final String environment,
...
+     * <p>
+     * Please note, the ICacheRecords returned by this query are NOT fully populated. Only the
+     * {@link GenericAccount} and {@link GenericRefreshToken} will be returned.
+     * will be resutned.
+     *
+     * @param environment The environment whose accounts should be queried.
+     * @return A List of ICacheRecords for the FoCI accounts.
+     */
+    @SuppressWarnings("unchecked")
+    public List<ICacheRecord> getFociCacheRecords(@NonNull final String environment) {
```
Comment (heidijinxujia) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java]: Do we need to set the `environment` `nonnull` here? In `BrokerOAuth2TokenCache#getAccounts(java.lang.String, java.lang.String)`, `environment` is nullable.

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java [hunk 2]
```diff
@@ -750,6 +753,81 @@ public AccountDeletionRecord removeAccount(@Nullable final String environment,
...
 
+    /**
+     * Tests if a clientId is 'known' to the cache. A clientId is known if a token has been
+     * previously saved to the cache with it.
+     *
+     * @param clientId The clientId whose known status should be queried.
+     * @return True if this clientId is known. False otherwise.
+     */
+    public boolean isClientIdKnownToCache(@NonNull final String clientId) {
+        return getAllClientIds().contains(clientId);
```
Comment (heidijinxujia) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java]: When will the known clientIds set get updated? By saving a new token?

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java [hunk 3]
```diff
@@ -750,6 +753,81 @@ public AccountDeletionRecord removeAccount(@Nullable final String environment,
...
+        final List<ICacheRecord> fociCacheRecords = new ArrayList<>();
+
+        final List<Pair<AccountRecord, String>> fociAccounts = new ArrayList<>();
+
+        final Set<String> fociClientIds = mApplicationMetadataCache.getAllFociClientIds();
+
+        for (final String fociClientId : fociClientIds) {
+            // Get the list of all of the accounts for this client...
+            final List<AccountRecord> clientAccounts = mFociCache.getAccounts(
+                    environment,
```
Comment (heidijinxujia) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java]: Checked. The `environment` here is nullable.

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java [hunk 4]
```diff
@@ -750,6 +753,81 @@ public AccountDeletionRecord removeAccount(@Nullable final String environment,
...
+        }
+
+        // For each account, load its refresh token and add it to the result List...
+        for (final Pair<AccountRecord, String> recordPair : fociAccounts) {
+            // Load the refresh token for this account...
+            final List<Credential> refreshTokens = mFociCache
+                    .getAccountCredentialCache()
+                    .getCredentialsFilteredBy(
+                            recordPair.first.getHomeAccountId(),
+                            environment,
```
Comment (heidijinxujia) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java]: Checked `SharedPreferencesAccountCredentialCache#getCredentialsFilteredBy`, the `environment` here is also nullable.

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java (hunk 2 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java]: Yes

**Pull request merged**